README.md
carbonmem: memory-backed carbon store
=================================================

<a href="https://packagecloud.io/"><img alt="Private Maven, RPM, DEB, PyPi and RubyGem Repository | packagecloud" height="46" src="https://packagecloud.io/images/packagecloud-badge.png" width="158" /></a>

Optimized to answer to topK kind of queries.

Other pieces of the stack are:
   - [carbonapi](https://github.com/go-graphite/carbonapi)
   - [carbonzipper](https://github.com/go-graphite/carbonzipper)
   - [carbonsearch](https://github.com/kanatohodets/carbonsearch)

For an overview of the stack and how the pieces fit together, watch
[Graphite@Scale or How to store millions metrics per second](https://fosdem.org/2017/schedule/event/graphite_at_scale/)
from FOSDEM '17.

Changes and versioning
----------------------

Version policy - all the versions we run in production is taged.

In case change will require simultanious upgrade of different components, it will be stated in Upgrading notes below.

Also we will try to maintain backward compatibility from down to up.

For example with protobuf2 -> protobuf3 migration - carbonzipper can still send results to older versions of carbonapi, but can't get results from older versions of carbonserver/go-carbon.

See [CHANGES.md](https://github.com/go-graphite/carbonmem/blob/master/CHANGES.md)


Work scheme
-----------

![carbonmem.png](doc/carbonmem.png)

Gray components are optional or alternative


Build
-----

Required golang 1.7+

```sh
# build binary
git clone https://github.com/go-graphite/carbonmem.git
cd carbonmem
make dep
make
```


Usage
-----

The service accept only following types of requests:

```toml
path.to.your.metrics.for.top.*.TopK.10m
```

The asterisk must be in metrics path. Usage of asterisk at any node of the tree are permitted.


Acknowledgement
---------------
This program was originally developed for Booking.com.  With approval
from Booking.com, the code was generalised and published as Open Source
on github, for which the author would like to express his gratitude.

License
-------

This code is licensed under the MIT license.


Contact
-------

If you have questions or problems there are two ways to contact us:

1. Open issue on a github page
2. #zipperstack on [gophers slack](https://invite.slack.golangbridge.org/)


CHANGES.md
Changes
=================================================

[Fix] - bugfix

**[Breaking]** - breaking change

[Feature] - new feature

[Improvement] - non-breaking improvement

[Code] - code quality related change that shouldn't make any significant difference for end-user

CHANGELOG
---------
**0.7**
 - [Fix] Make it compatible with both pb2 and pb3. Incompatibility was introduced in 0.6
 - [Fix] Make it compatible with recent carbonzipper definitions
 - [Code] commit vendored dependencies

**0.6.1**
 - [Fix] Fix panic on bad queries
 - [Improvement] Introduce vendoring and Makefile

**0.6**
 - **[Breaking]** Migrate to protobuf3

Notes on upgrading to 0.6.0:

You need to upgrade carbonzipper or go-carbon to the version that supports protobuf3

**<=0.5**
There is no dedicated changelog for older versions of carbonmem. Please see commit log for more information about what changed for each commit.


mem_test.go
package carbonmem

import (
	"math"
	"reflect"
	"testing"
)

// tests for memory-backed carbon store

func TestMemStore(t *testing.T) {

	var tests = []struct {
		metric   string
		t        int32
		count    uint64
		from     int32
		until    int32
		expected []float64
	}{
		{"foo", 120, 1, 119, 123, []float64{math.NaN(), 1}},
		{"foo", 120, 3, 119, 123, []float64{math.NaN(), 3}},
		{"foo", 121, 4, 119, 123, []float64{math.NaN(), 3, 4}},
		{"foo", 125, 4, 119, 125, []float64{math.NaN(), 3, 4, math.NaN(), math.NaN(), math.NaN(), 4}},
		{"foo", 129, 9, 125, 129, []float64{4, math.NaN(), math.NaN(), math.NaN(), 9}},
		{"foo", 124, 4, 120, 129, []float64{3, 4, math.NaN(), math.NaN(), 4, 4, math.NaN(), math.NaN(), math.NaN(), 9}},
		{"foo", 118, 12, 118, 129, []float64{24, math.NaN()}},
		{"foo", 130, 10, 0, 0, nil},
		{"foo", 131, 11, 0, 0, nil},
		{"foo", 128, 4, 127, 131, []float64{math.NaN(), 4, 9, 10, 11}},
		{"bar", 182, 11, 120, 240, []float64{math.NaN(), 11}},
		{"bar", 245, 22, 120, 240, []float64{11, 22}},
		{"bar", 305, 33, 120, 300, []float64{22, 33}},
	}

	w := NewWhisper(100, 10, 120, TrigramCutoff(1))

	var fooid MetricID = math.MaxUint32

	for _, tt := range tests {
		w.Set(tt.t, tt.metric, tt.count)
		if fooid == math.MaxUint32 {
			fooid, _ = w.l.Find("foo")
		}
		if tt.expected != nil {
			r := w.Fetch(tt.metric, tt.from, tt.until)
			if !nearlyEqual(r.Values, tt.expected) {
				t.Errorf("got %#v , want %#v\n", r.Values, tt.expected)
			}
		}
	}

	if _, ok := w.l.Find("foo"); ok {
		t.Errorf("foo not pruned from known keys")
	}

	w.Set(305, "baz", 1)
	bazid, ok := w.l.Find("baz")
	if !ok || bazid != fooid {
		t.Errorf("`baz` did not reuse fooid: got (%v,%v), want %v", bazid, ok, fooid)
	}
}

const eps = 0.0000000001

func nearlyEqual(a, b []float64) bool {
	if len(a) != len(b) {
		return false
	}

	for i, v := range a {
		// "same"
		if math.IsNaN(v) || math.IsNaN(b[i]) {
			if math.IsNaN(v) && math.IsNaN(b[i]) {
				continue
			}
			return false
		}
		// "close enough"
		if math.Abs(v-b[i]) > eps {
			return false
		}

	}

	return true
}

func TestGlob(t *testing.T) {

	w := NewWhisper(100, 10, 60)

	w.Set(100, "carbon.relays", 1)
	w.Set(100, "carbon.zipper", 1)
	w.Set(100, "carbon.rewhatever.errors", 1)
	w.Set(100, "carbon.rewhatever.count", 1)
	w.Set(100, "carbon.notmatched", 1)

	var tests = []struct {
		target string
		want   []Glob
	}{
		{
			"carbon.relays",
			[]Glob{
				{Metric: "carbon.relays", IsLeaf: true},
			},
		},
		{
			"carbon.re",
			nil,
		},
		{
			"carbon.re*",
			[]Glob{
				{Metric: "carbon.relays", IsLeaf: true},
				{Metric: "carbon.rewhatever", IsLeaf: false},
			},
		},
		{
			"carbon.z*",
			[]Glob{
				{Metric: "carbon.zipper", IsLeaf: true},
			},
		},
		{
			"carbon.z",
			nil,
		},
		{
			"carbon.re*.errors",
			[]Glob{
				{Metric: "carbon.rewhatever.errors", IsLeaf: true},
			},
		},
		{
			"carbon.re*eve*",
			[]Glob{
				{Metric: "carbon.rewhatever", IsLeaf: false},
			},
		},
		{
			"carbon.re*.erro*",
			[]Glob{
				{Metric: "carbon.rewhatever.errors", IsLeaf: true},
			},
		},
	}

	for _, tt := range tests {
		globs := w.Find(tt.target)
		if !reflect.DeepEqual(globs, tt.want) {
			t.Errorf("Find(%v)=%#v , want %#v\n", tt.target, globs, tt.want)
		}
	}
}

func TestTopK(t *testing.T) {

	w := NewWhisper(100, 60, 120)

	w.Set(100, "carbon.rewhatever.lots", 10)
	w.Set(100, "carbon.rewhatever.fewer", 8)
	w.Set(161, "carbon.rewhatever.foo", 5)
	w.Set(161, "carbon.rewhatever.bar", 4)
	w.Set(161, "carbon.notmatched", 1)

	var tests = []struct {
		prefix  string
		seconds int32
		globs   []Glob
	}{
		{"carbon.rewhatever.*", 60,
			[]Glob{
				{Metric: "carbon.rewhatever.foo", IsLeaf: true},
				{Metric: "carbon.rewhatever.bar", IsLeaf: true},
			},
		},
		{"carbon.rewhatever.*", 120,
			[]Glob{
				{Metric: "carbon.rewhatever.lots", IsLeaf: true},
				{Metric: "carbon.rewhatever.fewer", IsLeaf: true},
				{Metric: "carbon.rewhatever.foo", IsLeaf: true},
				{Metric: "carbon.rewhatever.bar", IsLeaf: true},
			},
		},
	}

	for _, tt := range tests {
		if g := w.TopK(tt.prefix, tt.seconds); !reflect.DeepEqual(g, tt.globs) {
			t.Errorf("Topk(%v, %v)=%v, want %v", tt.prefix, tt.seconds, g, tt.globs)
		}
	}
}


main_test.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"reflect"
	"sort"
	"testing"

	"github.com/gogo/protobuf/proto"

	"github.com/davecgh/go-spew/spew"
	"github.com/go-graphite/carbonmem/mwhisper"
	pb "github.com/go-graphite/carbonzipper/carbonzipperpb"
)

func TestParseTopK(t *testing.T) {

	var tests = []struct {
		query   string
		prefix  string
		seconds int32
		ok      bool
	}{
		// good
		{"prefix.blah.*.TopK.10m", "prefix.blah.*", 10 * 60, true},
		{"prefix.blah.*.TopK.10s", "prefix.blah.*", 10, true},
		{"prefix.blah.foo.TopK.10m", "prefix.blah.foo", 600, true},

		// bad
		{"prefix.foo.bar.baz", "", 0, false},
		{"prefix.blah.TopK.10m*", "", 0, false},
		{"prefix.blah.TopK.10", "", 0, false},
		{"prefix.blah.TopK.10m.", "", 0, false},
		{"prefix.blah.TopK.10z", "", 0, false},
		{"prefix.blah.TopK.10m.*.foo", "", 0, false},
		{"prefix.blah.TopK.10m.foo.*", "", 0, false},
		{"prefix.blah.TopK.10s*", "", 0, false},
		{"prefix.blah.TopK.-10m.*", "", 0, false},
		{"prefix.blah.TopK.*", "", 0, false},
		{"prefix.blah.TopK.", "", 0, false},
	}

	for _, tt := range tests {
		if prefix, seconds, ok := parseTopK(tt.query); prefix != tt.prefix || seconds != tt.seconds || ok != tt.ok {
			t.Errorf("parseTopK(%s)=(%q,%v,%v), want (%q,%v,%v)", tt.query, prefix, seconds, ok, tt.prefix, tt.seconds, tt.ok)
		}
	}
}

func TestTopKFind(t *testing.T) {

	metrics := mwhisper.NewWhisper(120, 60, 600)

	for _, m := range []struct {
		epoch  int32
		metric string
		count  uint64
	}{
		{120, "foo.bar", 10},
		{120, "foo.baz", 50},
		{120 + 1*60, "foo.bar", 11},
		{120 + 1*60, "foo.baz", 40},

		{120 + 2*60, "foo.bar", 12},
		{120 + 3*60, "foo.bar", 13},
		{120 + 3*60, "foo.qux", 13},
		{120 + 4*60, "foo.bar", 14},
	} {
		metrics.Set(m.epoch, m.metric, m.count)
	}

	Whispers.metrics = map[string]*mwhisper.Whisper{"foo": metrics}

	for _, tt := range []struct {
		query string
		want  pb.GlobResponse
	}{
		{
			"foo.*",
			pb.GlobResponse{
				Name: proto.String("foo.*"),
				Matches: []*pb.GlobMatch{
					&pb.GlobMatch{Path: proto.String("foo.bar"), IsLeaf: proto.Bool(true)},
					&pb.GlobMatch{Path: proto.String("foo.baz"), IsLeaf: proto.Bool(true)},
					&pb.GlobMatch{Path: proto.String("foo.qux"), IsLeaf: proto.Bool(true)},
				},
			},
		},

		{
			"foo.*.TopK.3m",
			pb.GlobResponse{
				Name: proto.String("foo.*.TopK.3m"),
				Matches: []*pb.GlobMatch{
					&pb.GlobMatch{Path: proto.String("foo.bar.TopK.3m"), IsLeaf: proto.Bool(true)},
					&pb.GlobMatch{Path: proto.String("foo.qux.TopK.3m"), IsLeaf: proto.Bool(true)},
				},
			},
		},

		{
			"foo.*.TopK.5m",
			pb.GlobResponse{
				Name: proto.String("foo.*.TopK.5m"),
				Matches: []*pb.GlobMatch{
					&pb.GlobMatch{Path: proto.String("foo.baz.TopK.5m"), IsLeaf: proto.Bool(true)},
					&pb.GlobMatch{Path: proto.String("foo.bar.TopK.5m"), IsLeaf: proto.Bool(true)},
					&pb.GlobMatch{Path: proto.String("foo.qux.TopK.5m"), IsLeaf: proto.Bool(true)},
				},
			},
		},
	} {
		req, _ := http.NewRequest("GET", fmt.Sprintf("/metrics/find/?query=%s&format=json", tt.query), nil)

		w := httptest.NewRecorder()

		findHandler(w, req)

		var response pb.GlobResponse

		json.Unmarshal(w.Body.Bytes(), &response)

		if !reflect.DeepEqual(response, tt.want) {
			t.Errorf("Find(%s)=%#v, want %#v", tt.query, spew.Sdump(response), spew.Sdump(tt.want))
		}
	}
}

func TestTopKRender(t *testing.T) {

	metrics := mwhisper.NewWhisper(100, 10, 60)

	for _, m := range []struct {
		epoch  int32
		metric string
		count  uint64
	}{
		{120, "foo.bar", 10},
		{120, "foo.baz", 50},
		{121, "foo.bar", 11},
		{121, "foo.baz", 40},

		{122, "foo.bar", 12},
		{123, "foo.bar", 13},
		{123, "foo.qux", 13},
		{124, "foo.bar", 14},
	} {
		metrics.Set(m.epoch, m.metric, m.count)
	}

	Whispers.metrics = map[string]*mwhisper.Whisper{"foo": metrics}

	for _, tt := range []struct {
		target string
		from   int32
		until  int32
		want   pb.MultiFetchResponse
	}{
		{
			"foo.bar",
			120, 124,
			pb.MultiFetchResponse{
				Metrics: []*pb.FetchResponse{
					&pb.FetchResponse{
						Name:      proto.String("foo.bar"),
						StartTime: proto.Int32(120),
						StopTime:  proto.Int32(124),
						StepTime:  proto.Int32(1),
						Values:    []float64{10, 11, 12, 13, 14},
						IsAbsent:  []bool{false, false, false, false, false},
					},
				},
			},
		},

		{
			"foo.bar.TopK.3s",
			120, 124,
			pb.MultiFetchResponse{
				Metrics: []*pb.FetchResponse{
					&pb.FetchResponse{
						Name:      proto.String("foo.bar.TopK.3s"),
						StartTime: proto.Int32(120),
						StopTime:  proto.Int32(124),
						StepTime:  proto.Int32(1),
						Values:    []float64{10, 11, 12, 13, 14},
						IsAbsent:  []bool{false, false, false, false, false},
					},
				},
			},
		},
	} {
		req, _ := http.NewRequest("GET", fmt.Sprintf("/render/?target=%s&from=%d&until=%d&format=json", tt.target, tt.from, tt.until), nil)

		w := httptest.NewRecorder()

		renderHandler(w, req)

		var response pb.MultiFetchResponse

		json.Unmarshal(w.Body.Bytes(), &response)

		if !reflect.DeepEqual(response, tt.want) {
			t.Errorf("Render(%s,%d,%d)=%#v, want %#v", tt.target, tt.from, tt.until, spew.Sdump(response), spew.Sdump(tt.want))
		}
	}
}

func TestNodePrefix(t *testing.T) {

	var tests = []struct {
		metric string
		n      int
		want   string
	}{
		{"foo.bar.baz.qux", 3, "foo.bar.baz"},
	}

	for _, tt := range tests {
		if got := findNodePrefix(tt.n, tt.metric); got != tt.want {
			t.Errorf("findNodePrefix(%d, %q)=%q, want %q", tt.n, tt.metric, got, tt.want)
		}
	}
}

func TestWhisperGlob(t *testing.T) {

	var whispers = whispers{
		metrics: map[string]*mwhisper.Whisper{
			"foo.bar.baz.qux": nil,
			"foo.bar.boz.qux": nil,
			"foo.bar.baz.zot": nil,
		},
	}

	for _, tt := range []struct {
		query string
		want  []string
	}{
		{
			"foo.*",
			[]string{
				"foo.bar",
				"foo.bar",
				"foo.bar",
			},
		},
		{
			"foo.bar.*",
			[]string{
				"foo.bar.baz",
				"foo.bar.baz",
				"foo.bar.boz",
			},
		},
		{
			"foo.bar.b*z.qux",
			[]string{
				"foo.bar.baz.qux",
				"foo.bar.boz.qux",
			},
		},
	} {

		got := whispers.Glob(tt.query)

		var strs []string
		for _, g := range got {
			strs = append(strs, g.Metric)
		}

		sort.Strings(strs)

		if !reflect.DeepEqual(strs, tt.want) {
			t.Errorf("Whispers.Glob(%s)=%v, want %v", tt.query, strs, tt.want)
		}
	}

}


main.go
// memory-backed carbon store: speaks graphite in, zipper out
// At the moment, shares a lot of code with grobian/carbonserver
package main

import (
	"bufio"
	"encoding/json"
	"errors"
	"expvar"
	"flag"
	"fmt"
	"math"
	"net"
	"net/http"
	_ "net/http/pprof"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	pb "github.com/go-graphite/carbonzipper/carbonzipperpb3"
	"github.com/go-graphite/carbonzipper/mlog"
	"github.com/go-graphite/carbonzipper/mstats"

	"github.com/peterbourgon/g2g"

	"github.com/go-graphite/carbonmem/mwhisper"
)

var BuildVersion = "(development build)"

var Metrics = struct {
	FindRequests  *expvar.Int
	FetchRequests *expvar.Int
}{
	FindRequests:  expvar.NewInt("findRequests"),
	FetchRequests: expvar.NewInt("fetchRequests"),
}
var logger mlog.Level

func parseTopK(query string) (string, int32, bool) {

	// prefix.blah.*.TopK.10m  => "prefix.blah.*", 600, true

	var idx int
	if idx = strings.Index(query, ".TopK."); idx == -1 {
		// not found
		return "", 0, false
	}

	prefix := query[:idx]

	timeIdx := idx + len(".TopK.")

	// look for number followed by 'm' or 's'
	unitsIdx := timeIdx
	for unitsIdx < len(query) && '0' <= query[unitsIdx] && query[unitsIdx] <= '9' {
		unitsIdx++
	}

	// ran off the end or no numbers present
	if unitsIdx == len(query) || unitsIdx == timeIdx {
		return "", 0, false
	}

	multiplier := 0
	switch query[unitsIdx] {
	case 's':
		multiplier = 1
	case 'm':
		multiplier = 60
	default:
		// unknown units
		return "", 0, false
	}

	if unitsIdx != len(query)-1 {
		return "", 0, false
	}

	timeUnits, err := strconv.Atoi(query[timeIdx:unitsIdx])
	if err != nil {
		return "", 0, false
	}

	return prefix, int32(timeUnits * multiplier), true
}

func findHandler(w http.ResponseWriter, req *http.Request) {

	Metrics.FindRequests.Add(1)

	query := req.FormValue("query")
	format := req.FormValue("format")

	if format != "json" && format != "protobuf3" && format != "protobuf" {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}

	matches := findMetrics(query)

	response := pb.GlobResponse{
		Name:    query,
		Matches: matches,
	}

	var b []byte
	switch format {
	case "json":
		w.Header().Set("Content-Type", "application/json")
		b, _ = json.Marshal(response)
	case "protobuf", "protobuf3":
		w.Header().Set("Content-Type", "application/protobuf")
		b, _ = response.Marshal()
	}
	w.Write(b)
}

func hasMetaCharacters(query string) bool {
	return strings.IndexByte(query, '*') != -1 || strings.IndexByte(query, '[') != -1 || strings.IndexByte(query, '?') != -1
}

func findMetrics(query string) []pb.GlobMatch {
	var topk string

	var globs []mwhisper.Glob

	if strings.Count(query, ".") < Whispers.prefix {
		globs = Whispers.Glob(query)
	} else {
		if m := Whispers.Fetch(query); m != nil {
			if prefix, seconds, ok := parseTopK(query); ok {
				topk = query[len(prefix):]
				if hasMetaCharacters(query) {
					globs = m.TopK(prefix, seconds)
				} else {
					globs = m.Find(prefix)
				}
			} else {
				globs = m.Find(query)
			}
		}
	}

	var matches []pb.GlobMatch
	paths := make(map[string]struct{}, len(globs))
	for _, g := range globs {
		// fix up metric name
		metric := g.Metric + topk
		if _, ok := paths[metric]; !ok {
			m := pb.GlobMatch{
				Path:   metric,
				IsLeaf: g.IsLeaf,
			}
			matches = append(matches, m)
			paths[metric] = struct{}{}
		}
	}

	return matches
}

func renderHandler(w http.ResponseWriter, req *http.Request) {

	Metrics.FetchRequests.Add(1)

	target := req.FormValue("target")
	format := req.FormValue("format")
	from := req.FormValue("from")
	until := req.FormValue("until")

	frint, _ := strconv.Atoi(from)
	unint, _ := strconv.Atoi(until)

	if unint < frint {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}

	if format != "json" && format != "protobuf3" && format != "protobuf" {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}

	matches := findMetrics(target)

	var multi pb.MultiFetchResponse

	for _, m := range matches {

		target := m.GetPath()

		var metric string
		if prefix, _, ok := parseTopK(target); ok {
			metric = prefix
		} else {
			metric = target
		}

		metrics := Whispers.Fetch(metric)
		if metrics == nil {
			continue
		}
		points := metrics.Fetch(metric, int32(frint), int32(unint))

		if points == nil {
			continue
		}

		fromTime := points.From
		untilTime := points.Until
		step := points.Step
		response := pb.FetchResponse{
			Name:      target,
			StartTime: fromTime,
			StopTime:  untilTime,
			StepTime:  step,
			Values:    make([]float64, len(points.Values)),
			IsAbsent:  make([]bool, len(points.Values)),
		}

		for i, p := range points.Values {
			if math.IsNaN(p) {
				response.Values[i] = 0
				response.IsAbsent[i] = true
			} else {
				response.Values[i] = p
				response.IsAbsent[i] = false
			}
		}

		multi.Metrics = append(multi.Metrics, response)
	}

	var b []byte
	switch format {
	case "json":
		w.Header().Set("Content-Type", "application/json")
		b, _ = json.Marshal(multi)
	case "protobuf", "protobuf3":
		w.Header().Set("Content-Type", "application/protobuf")
		b, _ = multi.Marshal()
	}
	w.Write(b)
}

func graphiteServer(port int) {

	ln, e := net.Listen("tcp", ":"+strconv.Itoa(port))

	if e != nil {
		logger.Fatalln("listen error:", e)
	}

	logger.Logln("graphite server starting on port", port)

	for {
		conn, err := ln.Accept()
		if err != nil {
			logger.Logln(err)
			continue
		}
		go func(c net.Conn) {
			scanner := bufio.NewScanner(c)
			for scanner.Scan() {
				metric, count, epoch, err := parseGraphite(scanner.Bytes())
				if err != nil {
					continue
				}

				metrics := Whispers.FetchOrCreate(metric)

				metrics.Set(int32(epoch), metric, uint64(count))
			}
			if err := scanner.Err(); err != nil {
				logger.Logf("graphite server: error during scan: %v", err)
			}
			c.Close()
		}(conn)
	}
}

func isspace(c byte) bool {
	return c == ' ' || c == '\t' || c == '\r' || c == '\n'
}

var errParseError = errors.New("graphite: parse error")

func token(b []byte) ([]byte, []byte) {

	if len(b) == 0 {
		return nil, nil
	}

	// munch space
	for len(b) > 0 && isspace(b[0]) {
		b = b[1:]
	}

	var i int
	for i < len(b) && !isspace(b[i]) {
		i++
	}

	return b[:i], b[i:]
}

func parseGraphite(b []byte) (metric string, count int, epoch int, err error) {

	var tok []byte

	tok, b = token(b)
	if len(tok) == 0 {
		return "", 0, 0, errParseError
	}

	metric = string(tok)

	tok, b = token(b)
	if len(tok) == 0 {
		return "", 0, 0, errParseError
	}

	count, err = strconv.Atoi(string(tok))
	if err != nil {
		return "", 0, 0, errParseError
	}

	tok, b = token(b)
	if len(tok) == 0 {
		return "", 0, 0, errParseError
	}

	epoch, err = strconv.Atoi(string(tok))
	if err != nil {
		return "", 0, 0, errParseError
	}

	// check for extra stuff
	tok, b = token(b)
	if len(tok) != 0 || len(b) != 0 {
		return "", 0, 0, errParseError
	}

	return metric, count, epoch, nil
}

type whispers struct {
	sync.RWMutex
	metrics map[string]*mwhisper.Whisper

	windowSize int
	epochSize  int
	epoch0     int
	prefix     int
}

var Whispers whispers

func findNodePrefix(prefix int, metric string) string {

	var found int
	for i, c := range metric {
		if c == '.' {
			found++
			if found >= prefix {
				return metric[:i]
			}
		}
	}
	return metric
}

func (w *whispers) FetchOrCreate(metric string) *mwhisper.Whisper {

	m := w.Fetch(metric)

	if m == nil {
		prefix := findNodePrefix(w.prefix, metric)
		var ok bool
		w.Lock()
		m, ok = w.metrics[prefix]
		if !ok {
			m = mwhisper.NewWhisper(int32(w.epoch0), w.epochSize, w.windowSize, mwhisper.TrigramCutoff(100000))
			w.metrics[prefix] = m
		}
		w.Unlock()
	}

	return m
}

func (w *whispers) Fetch(metric string) *mwhisper.Whisper {
	prefix := findNodePrefix(w.prefix, metric)

	w.RLock()
	m := w.metrics[prefix]
	w.RUnlock()

	return m
}

func (w *whispers) Glob(query string) []mwhisper.Glob {

	query = strings.Replace(query, ".", "/", -1)
	slashes := strings.Count(query, "/")

	w.RLock()
	var glob []mwhisper.Glob
	for m := range w.metrics {
		qm := strings.Replace(m, ".", "/", slashes)
		if trim := strings.Index(qm, "."); trim != -1 {
			qm = qm[:trim]
			m = m[:trim]
		}
		if match, err := filepath.Match(query, qm); err == nil && match {
			glob = append(glob, mwhisper.Glob{Metric: m})
		}
	}

	w.RUnlock()

	return glob
}

func main() {

	Whispers.metrics = make(map[string]*mwhisper.Whisper)

	flag.IntVar(&Whispers.windowSize, "w", 600, "window size")
	flag.IntVar(&Whispers.epochSize, "e", 60, "epoch window size")
	flag.IntVar(&Whispers.epoch0, "epoch0", 0, "epoch0")
	flag.IntVar(&Whispers.prefix, "prefix", 0, "prefix nodes to shard on")

	port := flag.Int("p", 8001, "port to listen on (http)")
	gport := flag.Int("gp", 2003, "port to listen on (graphite)")
	verbose := flag.Bool("v", false, "verbose logging")
	logdir := flag.String("logdir", "/var/log/carbonmem/", "logging directory")
	logtostdout := flag.Bool("stdout", false, "log also to stdout")
	graphiteHost := flag.String("graphite", "", "graphite destination host")
	interval := flag.Duration("interval", 60*time.Second, "interval to report internal statistics to graphite")

	flag.Parse()

	if *logdir == "" {
		mlog.SetRawStream(os.Stdout)
	} else {
		mlog.SetOutput(*logdir, "carbonmem", *logtostdout)
	}

	if *verbose {
		logger = mlog.Debug
	}

	expvar.NewString("BuildVersion").Set(BuildVersion)
	logger.Logln("starting carbonmem", BuildVersion)

	expvar.Publish("Whispers", expvar.Func(func() interface{} {
		m := make(map[string]int)
		Whispers.RLock()
		for k, v := range Whispers.metrics {
			m[k] = v.Len()
		}
		Whispers.RUnlock()
		return m
	}))

	if Whispers.epoch0 == 0 {
		Whispers.epoch0 = int(time.Now().Unix())
	}

	if envhost := os.Getenv("GRAPHITEHOST") + ":" + os.Getenv("GRAPHITEPORT"); envhost != ":" || *graphiteHost != "" {

		var host string

		switch {
		case envhost != ":" && *graphiteHost != "":
			host = *graphiteHost
		case envhost != ":":
			host = envhost
		case *graphiteHost != "":
			host = *graphiteHost
		}

		logger.Logln("Using graphite host", host)
		logger.Logln("setting stats interval to", *interval)

		// register our metrics with graphite
		graphite := g2g.NewGraphite(host, *interval, 10*time.Second)

		hostname, _ := os.Hostname()
		hostname = strings.Replace(hostname, ".", "_", -1)

		graphite.Register(fmt.Sprintf("carbon.mem.%s.find_requests", hostname), Metrics.FindRequests)
		graphite.Register(fmt.Sprintf("carbon.mem.%s.fetch_requests", hostname), Metrics.FetchRequests)

		go mstats.Start(*interval)

		graphite.Register(fmt.Sprintf("carbon.mem.%s.alloc", hostname), &mstats.Alloc)
		graphite.Register(fmt.Sprintf("carbon.mem.%s.total_alloc", hostname), &mstats.TotalAlloc)
		graphite.Register(fmt.Sprintf("carbon.mem.%s.num_gc", hostname), &mstats.NumGC)
		graphite.Register(fmt.Sprintf("carbon.mem.%s.pause_ns", hostname), &mstats.PauseNS)
	}

	go graphiteServer(*gport)

	http.HandleFunc("/metrics/find/", accessHandler(*verbose, findHandler))
	http.HandleFunc("/render/", accessHandler(*verbose, renderHandler))

	logger.Logln("http server starting on port", *port)
	logger.Fatalln(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
}

func accessHandler(verbose bool, handler http.HandlerFunc) http.HandlerFunc {
	if !verbose {
		return handler
	}
	return func(w http.ResponseWriter, r *http.Request) {
		t0 := time.Now()
		handler(w, r)
		since := time.Since(t0)
		logger.Logln(r.RequestURI, since.Nanoseconds()/int64(time.Millisecond))
	}
}
